### PR TITLE
fix: add node affinity for node where PVC is mounted to workspace's backup Jobs 

### LINKED ISSUE #1675: Backup/restore fails for per-user (common) storage on multi-node clusters due to PVC multi-attach
### Description

DevWorkspace backup/restore fails for workspaces using **per-user (common) storage** on multi-node clusters. The backup Job cannot mount the shared `ReadWriteOnce` PVC when it gets scheduled on a different node than the running workspace pod that holds the PVC.

This works correctly on single-node clusters (e.g., CRC) because all pods land on the same node, and Kubernetes allows multiple pods on the same node to mount a `ReadWriteOnce` PVC. On multi-node clusters (e.g., AWS with 6 nodes), the backup Job pod may be scheduled on a different node, causing a **Multi-Attach error** or the DWO controller skipping Job creation entirely.
```
  oc describe pod devworkspace-backup-wn85f-xkbgt -n dwo-backup-test

 ...

  Warning  FailedAttachVolume  29m  attachdetach-controller
  Multi-Attach error for volume
  "pvc-42803733-1cc7-4abb-a06c-4175bee35041" Volume is already used by
  pod(s) workspacec7ccc971e0c84d22-778bf568d4-5rzcj
```

**Per-workspace storage** is unaffected because each workspace has its own dedicated PVC — stopping one workspace fully releases its PVC for the backup Job.

### How To Reproduce

1. Deploy DWO on a multi-node OpenShift cluster (e.g., AWS with 6 nodes, OCP 4.20)
2. Create two DevWorkspaces using per-user (common) storage in the same namespace
3. Stop one workspace
4. Trigger a backup of the stopped workspace
5. Observe that the backup Job either:
   - Gets stuck in `Pending` with a Multi-Attach error (PVC is held by the running workspace on another node)
   - Is never created by DWO because the shared PVC is still in use
6. The backup times out

### Expected behavior

The backup Job should be scheduled on the same node as the running workspace pod (using pod affinity or nodeSelector targeting the node where the shared PVC is currently mounted), so that both can access the `ReadWriteOnce` PVC simultaneously.

### Additional context

- **Passes on CRC** (single-node): all pods share the same node, so RWO PVC mounting is not an issue
- **Fails on multi-node AWS clusters**: tested on OCP 4.20 with 6 worker nodes
- Per-workspace storage backup/restore works correctly on both single and multi-node clusters
- Tested during DWO pre-release testing (v0.42.0)
- Test results pattern:
  - ❌ Test 1-2: Per-User Storage + OCP Internal Registry — FAILED
  - ✅ Test 3-4: Per-Workspace Storage + OCP Internal Registry — PASSED
  - ❌ Test 5-6: Per-User Storage + External Registry (Quay) — FAILED
  - ✅ Test 7-8: Per-Workspace Storage + External Registry (Quay) — PASSED

**Suggested fix**: When creating a backup Job for per-user (common) storage, add a node affinity or pod affinity constraint to ensure the Job is scheduled on the same node where the shared PVC is currently mounted.

### PR DESCRIPTION
**Pull request by rohanKanojia**
### What does this PR do?
I noticed this issue while testing pre-release-testing claude workflow where Backup restore test scripts failed on clusterbot clusters.
> **Note:** The Multi-Attach error depends on Kubernetes scheduling the backup Job pod on a different node than where the PVC is currently mounted. This is non-deterministic: if both workspaces and the backup Job happen to land on the same node, the error won't occur even without the fix.

- On multi-node clusters, backup Jobs for workspaces using per-user (common) storage fail with `Multi-Attach error` when the Job is scheduled on a different node than the one where the ReadWriteOnce PVC is mounted
- Pin backup Jobs to the correct node using `NodeAffinity` with `kubernetes.io/hostname`, matching the existing pattern used by cleanup Jobs
- Consolidate duplicated node-lookup logic from `cleanup.go` and the backup controller into a shared `FindNodeForPVC` function in `pkg/provision/storage/shared.go`

### What issues does this PR fix or reference?
Fix #1675

By default, `claim-devworkspace` PVC uses `ReadWriteOnce` access mode, meaning it can only be attached to one node at a time. When another workspace is running and has the PVC mounted, the backup Job must be scheduled on the same node. Without node affinity, Kubernetes may schedule the Job on a different node, causing:

```
Warning  FailedAttachVolume  attachdetach-controller  Multi-Attach error for volume "pvc-xxx" Volume is already used by pod(s) workspace-xxx
```

The cleanup Job in `cleanup.go` already solved this by looking up which node has the PVC mounted and setting `NodeAffinity`. This PR:

1. Extracts the shared logic into `FindNodeForPVC(ctx, client, namespace, pvcName)` in `pkg/provision/storage/shared.go`
2. Refactors `cleanup.go` to use the shared function (removing ~37 lines of duplicated code)
3. Adds the same node affinity to backup Jobs in the `BackupCronJobReconciler`
4. Applies node affinity unconditionally for both per-user and per-workspace storage (harmless for per-workspace since there's no PVC contention)

### Is it tested? How?

Managed to reproduce the Multi-Attach error in a cluster bot cluster while running the backup/restore workflow.

<details><summary>Test case 1: NodeAffinity is added when another workspace is running (2 DevWorkspaces: 1 running, 1 stopped)</summary>

This verifies that when a running workspace has the PVC mounted, the backup Job for the stopped workspace gets NodeAffinity pinning it to the same node.

1. Deploy DWO from the `fix-backup-job-node-affinity` branch to the cluster

2. Enable backup in the `DevWorkspaceOperatorConfig`:
   ```bash
   oc patch devworkspaceoperatorconfig devworkspace-operator-config \
     -n openshift-operators --type=merge \
     -p '{"config":{"workspace":{"backupCronJob":{"enable":true,"schedule":"* * * * *","registry":{"path":"image-registry.openshift-image-registry.svc:5000"},"oras":{"extraArgs":"--insecure"}}}}}'
   ```

3. Create a test namespace:
   ```bash
   oc new-project dwo-backup-test
   ```

4. Create two DevWorkspaces that share the common PVC (`claim-devworkspace`):

   ```bash
   kubectl apply -f - <<EOF
   apiVersion: workspace.devfile.io/v1alpha2
   kind: DevWorkspace
   metadata:
     name: test-workspace-a
     namespace: dwo-backup-test
   spec:
     started: true
     template:
       projects:
         - name: web-nodejs-sample
           git:
             remotes:
               origin: "https://github.com/che-samples/web-nodejs-sample.git"
       components:
         - name: dev
           container:
             image: quay.io/devfile/universal-developer-image:latest
   EOF
   ```

   ```bash
   kubectl apply -f - <<EOF
   apiVersion: workspace.devfile.io/v1alpha2
   kind: DevWorkspace
   metadata:
     name: test-workspace-b
     namespace: dwo-backup-test
   spec:
     started: true
     template:
       projects:
         - name: web-nodejs-sample
           git:
             remotes:
               origin: "https://github.com/che-samples/web-nodejs-sample.git"
       components:
         - name: dev
           container:
             image: quay.io/devfile/universal-developer-image:latest
   EOF
   ```

   Both workspaces use the default `common` storage type, so they share the `claim-devworkspace` PVC.

5. Wait for both workspaces to reach `Running` state:
   ```bash
   oc get dw -n dwo-backup-test -w
   ```

6. Stop one of the workspaces (this triggers a backup Job on the next cron tick):
   ```bash
   oc patch dw test-workspace-a -n dwo-backup-test --type=merge \
     -p '{"spec":{"started":false}}'
   ```

7. Wait for the backup Job to be created (within 1 minute based on cron schedule):
   ```bash
   oc get jobs -n dwo-backup-test -w
   ```

8. Verify the backup Job has NodeAffinity pinning it to the node where `test-workspace-b` is running:
   ```bash
   oc get jobs -n dwo-backup-test -o yaml | grep -A 10 nodeAffinity
   ```
   Expected output:
   ```yaml
   nodeAffinity:
     requiredDuringSchedulingIgnoredDuringExecution:
       nodeSelectorTerms:
       - matchExpressions:
         - key: kubernetes.io/hostname
           operator: In
           values:
           - <node-where-test-workspace-b-is-running>
   ```

9. The backup Job pod should start successfully without Multi-Attach errors

</details>

<details><summary>Test case 2: NodeAffinity is not added when no other workspace is running (1 DevWorkspace, stopped)</summary>

This verifies that when no running workspace has the PVC mounted, the backup Job is created without NodeAffinity (since there's no contention risk).

1. Deploy DWO from the `fix-backup-job-node-affinity` branch to the cluster

2. Enable backup in the `DevWorkspaceOperatorConfig`:
   ```bash
   oc patch devworkspaceoperatorconfig devworkspace-operator-config \
     -n openshift-operators --type=merge \
     -p '{"config":{"workspace":{"backupCronJob":{"enable":true,"schedule":"* * * * *","registry":{"path":"image-registry.openshift-image-registry.svc:5000"},"oras":{"extraArgs":"--insecure"}}}}}'
   ```

3. Create a test namespace:
   ```bash
   oc new-project dwo-backup-test
   ```

4. Create a single DevWorkspace:

   ```bash
   kubectl apply -f - <<EOF
   apiVersion: workspace.devfile.io/v1alpha2
   kind: DevWorkspace
   metadata:
     name: test-workspace-a
     namespace: dwo-backup-test
   spec:
     started: true
     template:
       projects:
         - name: web-nodejs-sample
           git:
             remotes:
               origin: "https://github.com/che-samples/web-nodejs-sample.git"
       components:
         - name: dev
           container:
             image: quay.io/devfile/universal-developer-image:latest
   EOF
   ```

5. Wait for the workspace to reach `Running` state:
   ```bash
   oc get dw -n dwo-backup-test -w
   ```

6. Stop the workspace:
   ```bash
   oc patch dw test-workspace-a -n dwo-backup-test --type=merge \
     -p '{"spec":{"started":false}}'
   ```

7. Wait for the backup Job to be created (within 1 minute based on cron schedule):
   ```bash
   oc get jobs -n dwo-backup-test -w
   ```

8. Verify the backup Job does **not** have NodeAffinity (since no running pod mounts the PVC):
   ```bash
   oc get jobs -n dwo-backup-test -o yaml | grep -c nodeAffinity
   ```
   Expected output: `0`

9. The backup Job pod should complete successfully (no contention for the PVC)

</details>

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

## Summary by CodeRabbit

- **Bug Fixes**
  - Backup cron jobs now preferentially schedule on the node currently mounting the workspace’s PVC, improving backup reliability.
  - Storage cleanup jobs now use the same PVC-to-node detection, and automatically leave scheduling constraints unset when no matching running pod exists.

- **Tests**
  - Added unit tests to validate node-affinity behavior for both backup execution and PVC cleanup job generation, including matching, non-matching, and multiple-pod scenarios.

### CODE REVIEW

**Comment by tolusha** [comment]
Hi! I'm **che-ai-assistant** — I help with your pull requests.

**Available commands**:
- `/che-ai-assistant generate-che-doc` — Generate a documentation PR based on this PR's changes
- `/che-ai-assistant ok-pr-review` — Run a comprehensive PR review (summary, code review, deep review, impact analysis)
- `/che-ai-assistant ok-pr-readiness` — Ensure PR has validation steps
- `/che-ai-assistant check-pr-test-failures` — Analyze failing CI checks, identify root causes, and suggest fixes
- `/che-ai-assistant update-che-e2e-tests` — Update Eclipse Che e2e tests
- `/che-ai-assistant claude` — Run a free-form instruction on this PR
- `/che-ai-assistant help` — Show this help message

**Comment by tolusha** [comment]
/che-ai-assistant ok-pr-readiness


Task completed.

**Comment by tolusha** [comment]
## PR Readiness Assessment: PR #1677 — fix: add node affinity for node where PVC is mounted to workspace's backup Jobs

**Repository**: devfile/devworkspace-operator
**Linked Issue**: #1675 — Backup/restore fails for per-user (common) storage on multi-node clusters due to PVC multi-attach

---

| # | Criterion | Verdict | Notes |
|---|-----------|---------|-------|
| 1 | Problem Statement | ✅ PASS | Clearly describes the Multi-Attach error on multi-node clusters when backup Jobs are scheduled on a different node than where the ReadWriteOnce PVC is mounted. Both PR body and linked issue #1675 provide detailed context including actual error output. |
| 2 | Reproduction Steps | ✅ PASS | Issue #1675 provides a 6-step reproduction path. PR body includes two fully scripted test cases with exact `oc`/`kubectl` commands, resource manifests, and prerequisite configuration (multi-node cluster, per-user storage, two DevWorkspaces). |
| 3 | Expected Behavior After Fix | ✅ PASS | Clearly stated: backup Jobs get `NodeAffinity` pinning them to the node where a running pod mounts the PVC. Test case 1 shows expected YAML output with `kubernetes.io/hostname` match expression. Test case 2 confirms no affinity is set when no contention exists. |
| 4 | Scope of Changes | ✅ PASS | 6 files changed (+367/−55). All changes are logically related: extracts shared `FindNodeForPVC` and `NodeAffinityForHostname` into `shared.go`, refactors `cleanup.go` to use them (removing ~37 lines of duplication), adds node affinity to backup controller, and adds comprehensive tests. Rationale for each change is documented. |
| 5 | Test Evidence | ✅ PASS | New unit tests cover both positive and negative cases: backup controller tests (Job with/without node affinity), cleanup test (node affinity when pod mounts PVC), and table-driven tests for `FindNodeForPVC` (7 scenarios including running/pending pods, non-matching PVC, multiple pods, unlabeled pods). Manual testing confirmed on a clusterbot cluster. |
| 6 | Deployment & Verification Notes | ✅ PASS | Two detailed test cases provide step-by-step verification with exact commands: (1) verify NodeAffinity is added when another workspace runs on the PVC, (2) verify no NodeAffinity when no contention. Includes DWO deployment from the branch, backup config setup, workspace creation manifests, and specific `oc get` commands to inspect the result. |

**Overall**: ✅ READY

---

### Missing Information

- E2E test results are not yet available — the PR checklist shows `v8-devworkspace-operator-e2e` and `v8-che-happy-path` tests have not been triggered. Consider commenting `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to run them before merge.

### What's Good

- Excellent problem description with the actual Kubernetes error message and a clear explanation of why `ReadWriteOnce` PVC access mode causes this on multi-node clusters but not single-node
- The note about non-deterministic failure (depends on Kubernetes scheduling) is valuable context for reviewers
- Clean refactoring approach: extracts shared logic (`FindNodeForPVC`, `NodeAffinityForHostname`) rather than duplicating node-lookup code across backup and cleanup controllers
- Two well-structured manual test cases covering both the positive case (affinity added when PVC is contended) and the negative case (no affinity when no contention)
- Comprehensive table-driven unit tests for `FindNodeForPVC` covering 7 scenarios including edge cases (pending pods, unlabeled pods, non-matching PVC names, multiple pods)
- Linked issue #1675 provides additional reproduction context and test-result patterns across storage types and registry configurations

File: controllers/backupcronjob/backupcronjob_controller.go [hunk 1]
```diff
@@ -456,6 +457,32 @@ func (r *BackupCronJobReconciler) createBackupJob(
 			},
 		},
 	}
+	// Pin backup Job to the node where the PVC is currently mounted to avoid
+	// Multi-Attach errors with ReadWriteOnce PVCs on multi-node clusters.
+	targetNode, err := provstorage.FindNodeForPVC(ctx, r.Client, workspace.Namespace, pvc.Name)
+	if err != nil {
+		log.Error(err, "Failed to find node with PVC, backup Job will not have node affinity", "pvc", pvc.Name)
+	}
```
[the remark this answers is not in the document]
Comment (rohanKanojia) [reply controllers/backupcronjob/backupcronjob_controller.go:467]: The cleanup job (cleanup.go) already uses the cached clusterAPI.Client for the same purpose before this PR.

File: controllers/backupcronjob/backupcronjob_controller.go [hunk 2]
```diff
@@ -456,6 +457,32 @@ func (r *BackupCronJobReconciler) createBackupJob(
...
+									Operator: corev1.NodeSelectorOpIn,
+									Values:   []string{targetNode},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+	}
```
[the remark this answers is not in the document]
Comment (rohanKanojia) [reply controllers/backupcronjob/backupcronjob_controller.go:470]: The cleanup job uses the same pinning pattern without ActiveDeadlineSeconds either. This would be a new enhancement for both backup and cleanup, not a regression from this PR.

File: pkg/provision/storage/shared.go [hunk 3]
```diff
@@ -266,6 +268,60 @@ func getSharedPVCWorkspaceCount(namespace string, api sync.ClusterAPI) (total in
...
+		return "", err
+	}
+	podList := &corev1.PodList{}
+	if err := k8sClient.List(ctx, podList, &client.ListOptions{
+		Namespace:     namespace,
+		LabelSelector: labelSelector,
+	}); err != nil {
+		return "", err
+	}
+	return getNodeNameWithPVC(podList, pvcName), nil
```
[the remark this answers is not in the document]
Comment (rohanKanojia) [reply pkg/provision/storage/shared.go:286]: If the API call fails, it's better to create the backup Job without affinity (it might land on the right node) than to block backups entirely. Same pattern the cleanup code uses. Failing hard on a transient API error would be worse.

File: pkg/provision/storage/shared.go [hunk 4]
```diff
@@ -266,6 +268,60 @@ func getSharedPVCWorkspaceCount(namespace string, api sync.ClusterAPI) (total in
...
+								Operator: corev1.NodeSelectorOpIn,
+								Values:   []string{nodeName},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
```
[the remark this answers is not in the document]
Comment (rohanKanojia) [reply pkg/provision/storage/shared.go:310]: the cleanupjob has used this same NodeName → kubernetes.io/hostname pattern for a long time without issues.
On OpenShift (DWO's target), they match. Fixing this would be a change to both cleanup and backup, not scoped to this PR.

File: controllers/backupcronjob/backupcronjob_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkwon17) [reply controllers/backupcronjob/backupcronjob_controller.go:467]: I'm concerned that it may be a bit expensive to do a non caching list request for backup pods. If there are 1000 backup jobs to be created, there will be 1000 list requests.

**Comment by dkwon17** [comment]
LGTM, thank you for the detailed test instructions: 

Test case 1:
image


Test case 2:
image

**Review by dkwon17 — APPROVED** [review]
